pennylane/templates/subroutines/select.py

```diff
@@ -0,0 +1,162 @@
+    >>> @qml.qnode(dev)
+    >>> def circuit():
+    >>>     qml.Select(ops,control_wires=[0,1])
+    >>>     return qml.state()
+    ...
+    >>> print(qml.draw(circuit,expansion_strategy='device')())
+    0: ─╭○─╭○─╭●─╭●────┤  State
+    1: ─├○─├●─├○─├●────┤  State
+    2: ─╰X─│──╰Y─├SWAP─┤  State
+    3: ────╰X────╰SWAP─┤  State
+
+    """
+
+    num_wires = qml.operation.AnyWires
+
+    def _flatten(self):
+        return (self.ops), (self.control_wires)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "Select":
+        return cls(data, metadata)
+
+    def __init__(self, ops, control_wires, id=None):
+        control_wires = qml.wires.Wires(control_wires)
+        self.hyperparameters["ops"] = ops
+        self.hyperparameters["control_wires"] = control_wires
+
+        if 2 ** len(control_wires) < len(ops):
+            raise ValueError(
+                f"Not enough control wires ({len(control_wires)}) for the desired number of "
+                + f"operations ({len(ops)}). At least {int(math.ceil(math.log2(len(ops))))} control "
+                + "wires required."
+            )
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control_wires
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        for op in ops:
+            qml.QueuingManager.remove(op)
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+        all_wires = target_wires + control_wires
+        super().__init__(ops, wires=all_wires, id=id)
+
+    def decomposition(self):
+        r"""Representation of the operator as a product of other operators.
+
+        .. math:: O = O_1 O_2 \dots O_n
+
+        A ``DecompositionUndefinedError`` is raised if no representation by decomposition is defined.
+
+        .. seealso:: :meth:`~.Operator.compute_decomposition`.
+
+        Returns:
+            list[Operator]: decomposition of the operator
+        """
+        return self.compute_decomposition(self.ops, control_wires=self.control_wires)
+
+    @staticmethod
+    def compute_decomposition(
+        ops,
+        control_wires,
+    ):  # pylint: disable=arguments-differ, unused-argument
+        r"""Representation of the operator as a product of other operators (static method).
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+        .. note::
+
+            Operations making up the decomposition should be queued within the
+            ``compute_decomposition`` method.
+
+        .. seealso:: :meth:`~.Operator.decomposition`.
+
+        Args:
+            ops (list[Operator]): operations to apply
+            control_wires (Sequence[int]): the wires controlling which operation is applied
+
+        Returns:
+            list[Operator]: decomposition of the operator
+        """
+        states = [
```

Comment:
Not using `itertools.product` might lead to performance degradation
<img width="1203" alt="image" src="https://github.com/PennyLaneAI/pennylane/assets/23234279/18fe9aa5-b023-458d-a5b8-881bd7696313">

Reply:
Oh no. Back to `itertools` then.